File: ports/cpp/src/core/mac.cpp
#include "mac.hpp"

#include <cctype>
#include <format>

namespace pf {

static bool is_hex(char c) {
    return std::isxdigit(static_cast<unsigned char>(c)) != 0;
}

std::optional<Mac> parse_mac(std::string_view s) {
    Mac out{};
    size_t part = 0;
    size_t i = 0;
    while (part < 6) {
        if (i + 1 >= s.size())
            return std::nullopt;
        char a = s[i];
        char b = s[i + 1];
        if (!is_hex(a) || !is_hex(b))
            return std::nullopt;

        auto val = [](char c) -> int {
            if (c >= '0' && c <= '9') return c - '0';
            if (c >= 'a' && c <= 'f') return c - 'a' + 10;
            return c - 'A' + 10;
        };
        out[part] = static_cast<uint8_t>(val(a) * 16 + val(b));
        i += 2;
        ++part;

        if (part < 6) {
            if (i >= s.size() || s[i] != ':')
                return std::nullopt;
            ++i;
        }
    }
    if (i != s.size())
        return std::nullopt;
    return out;
}

std::string format_mac_bytes(const Mac& mac) {
    return std::format("{:02X}:{:02X}:{:02X}:{:02X}:{:02X}:{:02X}", mac[0], mac[1],
                       mac[2], mac[3], mac[4], mac[5]);
}

std::string format_mac_from_hex(std::string_view raw) {
    std::string hex;
    for (char c : raw) {
        if (is_hex(c)) {
            hex.push_back(static_cast<char>(std::toupper(static_cast<unsigned char>(c))));
            if (hex.size() == 12)
                break;
        }
    }
    std::string out;
    for (size_t i = 0; i < hex.size(); i += 2) {
        if (i > 0)
            out.push_back(':');
        out.push_back(hex[i]);
        if (i + 1 < hex.size())
            out.push_back(hex[i + 1]);
    }
    return out;
}

} // namespace pf

File: ports/zig/src/mac.zig
//! MAC address parsing / formatting.
//! Port of `parse_mac`, `format_mac_bytes` and `format_mac_from_hex`.
const std = @import("std");

pub const Mac = [6]u8;

fn hexVal(c: u8) ?u8 {
    return switch (c) {
        '0'...'9' => c - '0',
        'a'...'f' => c - 'a' + 10,
        'A'...'F' => c - 'A' + 10,
        else => null,
    };
}

/// Parse "AA:BB:CC:DD:EE:FF" into 6 bytes; null on any malformed input.
pub fn parseMac(s: []const u8) ?Mac {
    var out: Mac = undefined;
    var i: usize = 0;
    var part: usize = 0;
    while (part < 6) : (part += 1) {
        if (i + 1 >= s.len) return null;
        const hi = hexVal(s[i]) orelse return null;
        const lo = hexVal(s[i + 1]) orelse return null;
        out[part] = hi * 16 + lo;
        i += 2;
        if (part != 5) {
            if (i >= s.len or s[i] != ':') return null;
            i += 1;
        }
    }
    if (i != s.len) return null; // reject trailing garbage
    return out;
}

const HEX = "0123456789ABCDEF";

/// Format 6 bytes as "AA:BB:CC:DD:EE:FF" into `buf`, returning the used slice.
pub fn formatMacBytes(mac: Mac, buf: *[17]u8) []const u8 {
    var j: usize = 0;
    for (mac, 0..) |byte, idx| {
        if (idx > 0) {
            buf[j] = ':';
            j += 1;
        }
        buf[j] = HEX[byte >> 4];
        buf[j + 1] = HEX[byte & 0x0F];
        j += 2;
    }
    return buf[0..j];
}

/// Keep only hex digits (max 12), uppercase, regroup into colon-separated pairs.
pub fn formatMacFromHex(raw: []const u8, buf: *[17]u8) []const u8 {
    var hex: [12]u8 = undefined;
    var n: usize = 0;
    for (raw) |c| {
        if (std.ascii.isHex(c)) {
            hex[n] = std.ascii.toUpper(c);
            n += 1;
            if (n == 12) break;
        }
    }
    var j: usize = 0;
    var i: usize = 0;
    while (i < n) : (i += 2) {
        if (i > 0) {
            buf[j] = ':';
            j += 1;
        }
        buf[j] = hex[i];
        j += 1;
        if (i + 1 < n) {
            buf[j] = hex[i + 1];
            j += 1;
        }
    }
    return buf[0..j];
}

test "parseMac accepts well-formed addresses" {
    const m = parseMac("AA:BB:CC:DD:EE:FF").?;
    try std.testing.expectEqual(@as(u8, 0xAA), m[0]);
    try std.testing.expectEqual(@as(u8, 0xFF), m[5]);
    try std.testing.expect(parseMac("aa:bb:cc:dd:ee:ff") != null);
}

test "parseMac rejects malformed addresses" {
    try std.testing.expect(parseMac("AA:BB:CC:DD:EE") == null);
    try std.testing.expect(parseMac("AA:BB:CC:DD:EE:FF:00") == null);
    try std.testing.expect(parseMac("GG:BB:CC:DD:EE:FF") == null);
    try std.testing.expect(parseMac("AABBCCDDEEFF") == null);
}

test "formatMacBytes / formatMacFromHex" {
    var buf: [17]u8 = undefined;
    const bytes = Mac{ 0x01, 0x23, 0x45, 0x67, 0x89, 0xAB };
    try std.testing.expectEqualStrings("01:23:45:67:89:AB", formatMacBytes(bytes, &buf));
    try std.testing.expectEqualStrings("01:23:45:67:89:AB", formatMacFromHex("0123456789ab", &buf));
    try std.testing.expectEqualStrings("AA:BB:CC", formatMacFromHex("aa-bb-cc", &buf));
    try std.testing.expectEqualStrings("01:23:45:67:89:AB", formatMacFromHex("0123456789abCDEF99", &buf));
}
